server/modules/routing/dbseerroute/dbseerroute.c

```diff
@@ -0,0 +1,1181 @@
+static void log_closed_session(mysql_server_cmd_t mysql_command, bool is_closed,
+                               SERVER_REF *ref)
+{
+    char msg[MAX_SERVER_NAME_LEN + 200] = ""; // Extra space for message
+
+    if (is_closed)
+    {
+        sprintf(msg, "Session is closed.");
+    }
+    else if (SERVER_IS_DOWN(ref->server))
+    {
+        sprintf(msg, "Server '%s' is down.", ref->server->unique_name);
+    }
+    else if (!SERVER_REF_IS_ACTIVE(ref))
+    {
+        sprintf(msg, "Server '%s' was removed from the service.", ref->server->unique_name);
+    }
+
+    MXS_ERROR("Failed to route MySQL command %d to backend server. %s",
+              mysql_command, msg);
+}
+
+/**
+ * We have data from the client, we must route it to the backend.
+ * This is simply a case of sending it to the connection that was
+ * chosen when we started the client session.
+ *
+ * @param instance		The router instance
+ * @param router_session	The router session returned from the newSession call
+ * @param queue			The queue of data buffers to route
+ * @return if succeed 1, otherwise 0
+ */
+static int
+routeQuery(ROUTER *instance, void *router_session, GWBUF *queue)
+{
+    ROUTER_INSTANCE *inst = (ROUTER_INSTANCE *) instance;
+    ROUTER_CLIENT_SES *router_cli_ses = (ROUTER_CLIENT_SES *) router_session;
+    int rc = 0;
+    DCB* backend_dcb;
+    MySQLProtocol *proto = (MySQLProtocol*)router_cli_ses->client_dcb->protocol;
+    mysql_server_cmd_t mysql_command = proto->current_command;
+    bool rses_is_closed;
+
+    inst->stats.n_queries++;
+
+    /** Dirty read for quick check if router is closed. */
+    if (router_cli_ses->rses_closed)
+    {
+        rses_is_closed = true;
+    }
+    else
+    {
+        /**
+         * Lock router client session for secure read of DCBs
+         */
+        rses_is_closed = !(rses_begin_locked_router_action(router_cli_ses));
+    }
+
+    if (!rses_is_closed)
+    {
+        backend_dcb = router_cli_ses->backend_dcb;
+        /** unlock */
+        rses_end_locked_router_action(router_cli_ses);
+    }
+
+    if (rses_is_closed || backend_dcb == NULL ||
+        !SERVER_REF_IS_ACTIVE(router_cli_ses->backend) ||
+        SERVER_IS_DOWN(router_cli_ses->backend->server))
+    {
+        log_closed_session(mysql_command, rses_is_closed, router_cli_ses->backend);
+        gwbuf_free(queue);
+        goto return_rc;
+
+    }
+
+    char *ptr = NULL;
+    ptr = modutil_get_SQL(queue);
+
+    /* do performance logging. */
+    if (ptr && inst->log_filename)
+    {
+        router_cli_ses->sql_end = false;
+        int query_len = strlen(ptr);
+        /* check for commit and rollback */
+        if (query_len > 5)
+        {
+             int query_size = strlen(ptr)+1;
+             char *buf = router_cli_ses->buf;
+             for (int i = 0; i < query_size && i < buf_size; ++i)
+             {
+                 buf[i] = tolower(ptr[i]);
+             }
+             if (strncmp(buf, "commit", 6) == 0)
```

Comment:
You can use the query classifier for this. See `qc_get_type` in `query_classifier.h`.